force_bdss/data_sources/base_data_source_model.py

```diff
@@ -46,43 +47,187 @@ class BaseDataSourceModel(BaseModel):
 
         try:
             input_slots, output_slots = data_source.slots(self)
         except Exception:
             logger.exception(
                 "Unable to retrieve slot information from data source"
-                " created by factory '%s', plugin '%s'. This might "
+                " created by factory {}, plugin {}. This might "
                 "indicate a programming error.",
                 self.factory.id,
                 self.factory.plugin.id
             )
             raise
 
+        return input_slots, output_slots
+
+    def _assign_slot_info(self, name, new_slot_info):
+        """Assign either input_slot_info or output_slot_info attributes
+        with new values. The argument slot_info must have the same length as
+        the existing class trait name.
+
+        Parameters
+        ----------
+        name: str
+            Name of attribute to update, must be either "input_slot_info"
+            or "output_slot_info"
+        new_slot_info: list of type InputSlotInfo of OutputSlotInfo
+            List containing objects to update each element of
+            name attribute with
+
+        Raises
+        ------
+        RuntimeError, if length of slot_info and name attribute are not equal,
+        or if the `type` and `description` attributes on each element do not
+        match
+        """
+
+        # Ignore any changes if new_slot_info list is empty
+        if len(new_slot_info) == 0:
+            return
+
+        if name not in ["input_slot_info", "output_slot_info"]:
+            error_msg = (
+                "Attribute 'name' must be either 'input_slot_info' "
+                "or 'output_slot_info'."
+            )
+            logger.exception(error_msg)
+            raise RuntimeError(error_msg)
+
+        slot_info_attr = getattr(self, name)
+
+        if len(new_slot_info) != len(slot_info_attr):
+            error_msg = (
+                "The number of slots in {} ({}) of the {} model doesn't"
+                " match the expected number of slots ({}). This is"
+                " likely due to a corrupted file.".format(
+                    name, len(slot_info_attr),
+                    type(self).__name__,
+                    len(new_slot_info))
+            )
+            logger.exception(error_msg)
+            raise RuntimeError(error_msg)
+
+        # Check whether each element in new_slot_info has the same class,
+        # `type` and `description` attributes as the corresponding element
+        # in slot_info_attr, and if so, synchronize their `name` attributes
+        for new_info, old_info in zip(new_slot_info, slot_info_attr):
+            sync_trait_with_check(
+                new_info, old_info, 'name',
+                attr_checks=['__class__', 'type', 'description'],
+                ignore_default=True
+            )
+
+    # -------------------
+    #   Public Methods
+    # -------------------
+
+    def slot_info_defaults(self):
```

Comment:
Here we just provide a public method to obtain the default values of `input_slot_info` and `output_slot_info`, as they will be needed in `force_wfanager` after a `change_slot` event occurs